<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.28">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Reference Book - Beta version - PIgLET - R pakcage</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./cluster_docker.html" rel="next">
<link href="./allele_based_genotype.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="style.css">
<link rel="stylesheet" href="index.css">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">PIgLET - R pakcage</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./" class="sidebar-logo-link">
      <img src="./figures/myimage.gif" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Reference Book - Beta version</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/yaarilab/IGHV_reference_book/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Share" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-share"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
            <i class="bi bi-bi-twitter pe-1"></i>
          Twitter
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
            <i class="bi bi-bi-facebook pe-1"></i>
          Facebook
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.linkedin.com/sharing/share-offsite/?url=|url|">
            <i class="bi bi-bi-linkedin pe-1"></i>
          LinkedIn
          </a>
        </li>
    </ul>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preface.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./alleles_groups.html" class="sidebar-item-text sidebar-link">Allele Similarity Cluster (ASC)</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">Datasets and processing</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./allele_based_genotype.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">ASC-based genotype {.unnumbered}</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./piglet_package.html" class="sidebar-item-text sidebar-link active">PIgLET - R pakcage</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cluster_docker.html" class="sidebar-item-text sidebar-link">Processing AIRR-seq data with the ASC-pipeline</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Allele Similarity Clusters</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-G1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">IGHV1-18</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-G2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">IGHV1-2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-G3.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">IGHV1-24</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-G4.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">IGHV1-3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-G5.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">IGHV1-45</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-G6.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">IGHV1-46</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-G7.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">IGHV1-58</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-G8.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">IGHV1-69-2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-G9.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">IGHV1-69D/IGHV1-69</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-G10.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">IGHV1-8</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-G11.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">IGHV2-26</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-G12.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">IGHV2-5</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-G13.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">IGHV2-70D/IGHV2-70</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-G14.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">IGHV3-11</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15-G15.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">IGHV3-13</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16-G16.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">IGHV3-15</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17-G17.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">IGHV3-20</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./18-G18.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">IGHV3-21</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19-G19.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">IGHV3-23D/IGHV3-23</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20-G20.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">IGHV3-30-5/IGHV3-30/IGHV3-33/IGHV3-30-3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21-G21.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">IGHV3-35</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./22-G22.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">IGHV3-43D/IGHV3-43</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23-G23.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">IGHV3-48</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./24-G24.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">IGHV3-49</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./25-G25.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">IGHV3-66/IGHV3-53</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./26-G26.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">IGHV3-64D/IGHV3-64</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./27-G27.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">IGHV3-7</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./28-G28.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">IGHV3-73</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./29-G29.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">IGHV3-72</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./30-G30.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">IGHV3-74</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./31-G31.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">IGHV3-9</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./32-G32.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">IGHV4-28</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./33-G33.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">IGHV4-30-4/IGHV4-30-2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./34-G34.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">IGHV4-31</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./35-G35.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">IGHV4-34</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./36-G36.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">IGHV4-39/IGHV4-38-2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./37-G37.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">38</span>&nbsp; <span class="chapter-title">IGHV4-4/IGHV4-59</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./38-G38.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">IGHV4-61</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./39-G39.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">40</span>&nbsp; <span class="chapter-title">IGHV5-10-1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40-G40.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">41</span>&nbsp; <span class="chapter-title">IGHV5-51</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./41-G41.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">42</span>&nbsp; <span class="chapter-title">IGHV6-1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./42-G42.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">43</span>&nbsp; <span class="chapter-title">IGHV7-4-1</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#installing-piglet" id="toc-installing-piglet" class="nav-link active" data-scroll-target="#installing-piglet">Installing PIgLET</a></li>
  <li><a href="#allele-similarity-cluster" id="toc-allele-similarity-cluster" class="nav-link" data-scroll-target="#allele-similarity-cluster">Allele Similarity Cluster</a>
  <ul class="collapse">
  <li><a href="#sec-ASC" id="toc-sec-ASC" class="nav-link" data-scroll-target="#sec-ASC">Inferring Allele Similarity Clusters</a></li>
  <li><a href="#artificial-framework-1-reference-set" id="toc-artificial-framework-1-reference-set" class="nav-link" data-scroll-target="#artificial-framework-1-reference-set">Artificial framework 1 reference set</a></li>
  </ul></li>
  <li><a href="#allele-based-genotype" id="toc-allele-based-genotype" class="nav-link" data-scroll-target="#allele-based-genotype">Allele based genotype</a>
  <ul class="collapse">
  <li><a href="#asc-based-thresholds" id="toc-asc-based-thresholds" class="nav-link" data-scroll-target="#asc-based-thresholds">ASC-based thresholds</a></li>
  <li><a href="#inferring-asc-based-genotype" id="toc-inferring-asc-based-genotype" class="nav-link" data-scroll-target="#inferring-asc-based-genotype">Inferring ASC-based genotype</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/yaarilab/IGHV_reference_book/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">PIgLET - R pakcage</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="cell">

</div>
<p>Program for Ig clusters R package (PIgLET) is a suite of computational tools that improves genotype inference and downstream AIRR-seq data analysis. The package as two main tools. The first is Allele Clusters, this tool is designed to reduce the ambiguity within the IGHV alleles. The ambiguity is caused by duplicated or similar alleles which are shared among different genes. The second tool is an allele based genotype, that determined the presence of an allele based on a threshold derived from a naive population.</p>
<p>(figures/piglet_logo.svg)</p>
<section id="installing-piglet" class="level2">
<h2 class="anchored" data-anchor-id="installing-piglet">Installing PIgLET</h2>
<p>PIgLET is available for installation from the development version.</p>
<p>To build from the source code, first install the build dependencies:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"devtools"</span>, <span class="st">"roxygen2"</span>, <span class="st">"testthat"</span>, <span class="st">"knitr"</span>, <span class="st">"rmarkdown"</span>, <span class="st">"plotly"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To install the latest version via devtools:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install_bitbucket</span>(<span class="st">"yaarilab/piglet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note, installing from bitbucket does not generate the documentation. To generate them, first clone the repository and then build:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install_deps</span>()</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">document</span>()</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">build</span>()</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">install</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="allele-similarity-cluster" class="level2">
<h2 class="anchored" data-anchor-id="allele-similarity-cluster">Allele Similarity Cluster</h2>
<p>The functions in this section support the creation of the allele similarity cluster form an IGHV germline set, the table below shows the functions and a short description for each:</p>
<table class="table">
<colgroup>
<col style="width: 26%">
<col style="width: 73%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>inferAlleleClusters</td>
<td>The main function of the section to create the allele clusters based on a germline set</td>
</tr>
<tr class="even">
<td>ighvDistance</td>
<td>Calculate the distance between IGHV aligned germline sequences</td>
</tr>
<tr class="odd">
<td>ighvClust</td>
<td>Hierarchical clustering of the distance matrix from <code>ighvDistance</code></td>
</tr>
<tr class="even">
<td>generateReferenceSet</td>
<td>Generate the allele clusters reference set</td>
</tr>
<tr class="odd">
<td>plotAlleleCluster</td>
<td>Plots the Hierarchical clustering</td>
</tr>
<tr class="even">
<td>artificialFRW1Germline</td>
<td>Artificially create an IGHV reference set with framework1 (FWR1) primers</td>
</tr>
</tbody>
</table>
<section id="sec-ASC" class="level3">
<h3 class="anchored" data-anchor-id="sec-ASC">Inferring Allele Similarity Clusters</h3>
<p>The main function in this section <code>inferAlleleClusters</code> returns an S4 object that includes the ASC allele cluster table <code>alleleClusterTable</code> with the new names and the default thresholds, the renamed germline set <code>alleleClusterSet</code>, and the germline set hierarchical clustering <code>hclustAlleleCluster</code>, and the similarity threshold parameters <code>threshold</code>. Further by using the plot function on the returned object, a colorful visualization of the allele clusters dendrogram and threshold is received.</p>
<p>The function receives as an input a germline reference set of allele sequences, the filtration parameters for the 3’ and 5’ regions, and two similarity thresholds for the ASC clusters and families.</p>
<p>To create the clusters we will first load data from the package:</p>
<ol type="1">
<li>The IGHV germline reference - this reference set was download from IMGT in July 2022.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(piglet)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(HVGERM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>2.The allele functionality table - the table contains functionality information for each of the alleles. Download from IMGT in July 2022</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(hv_functionality)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before clustering the germline set, we will remove non functional alleles, alleles that do not start on the first 5’ nucleotide, and those that are shorter than 318 bases.</p>
<div class="cell">

</div>
<div id="lst-germlineset" class="listing">
<p>code listing 1: Clean germline set</p>
<div class="sourceCode" id="lst-germlineset" data-lst-cap="Clean germline set"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-germlineset-1"><a href="#lst-germlineset-1" aria-hidden="true" tabindex="-1"></a>germline <span class="ot">&lt;-</span> HVGERM</span>
<span id="lst-germlineset-2"><a href="#lst-germlineset-2" aria-hidden="true" tabindex="-1"></a><span class="do">## keep only functional alleles</span></span>
<span id="lst-germlineset-3"><a href="#lst-germlineset-3" aria-hidden="true" tabindex="-1"></a>germline <span class="ot">&lt;-</span> germline[hv_functionality<span class="sc">$</span>allele[hv_functionality<span class="sc">$</span>functional<span class="sc">==</span><span class="st">"F"</span>]]</span>
<span id="lst-germlineset-4"><a href="#lst-germlineset-4" aria-hidden="true" tabindex="-1"></a><span class="do">## keep only alleles that start from the first position of the V sequence</span></span>
<span id="lst-germlineset-5"><a href="#lst-germlineset-5" aria-hidden="true" tabindex="-1"></a>germline <span class="ot">&lt;-</span> germline[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"^[.]"</span>, germline)]</span>
<span id="lst-germlineset-6"><a href="#lst-germlineset-6" aria-hidden="true" tabindex="-1"></a><span class="do">## keep only alleles that are at minimum 318 nucleotide long</span></span>
<span id="lst-germlineset-7"><a href="#lst-germlineset-7" aria-hidden="true" tabindex="-1"></a>germline <span class="ot">&lt;-</span> germline[<span class="fu">nchar</span>(germline) <span class="sc">&gt;=</span> <span class="dv">318</span>]</span>
<span id="lst-germlineset-8"><a href="#lst-germlineset-8" aria-hidden="true" tabindex="-1"></a><span class="do">## keep only localized alleles (remove NL)</span></span>
<span id="lst-germlineset-9"><a href="#lst-germlineset-9" aria-hidden="true" tabindex="-1"></a>germline <span class="ot">&lt;-</span> germline[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"NL"</span>, <span class="fu">names</span>(germline))]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we will create the ASC clusters using the <code>inferAlleleClusters</code> function. For better clustering results with the human IGHV reference set, it is recommended to set the <code>trim_3prime_side</code> parameter to 318. Here, we will use the default similarity thresholds 75% for the family and 95% for the clusters.</p>
<div class="cell">

</div>
<div id="lst-asc" class="listing">
<p>code listing 2: ASC inference</p>
<div class="sourceCode" id="lst-asc" data-lst-cap="ASC inference"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-asc-1"><a href="#lst-asc-1" aria-hidden="true" tabindex="-1"></a>asc <span class="ot">&lt;-</span> <span class="fu">inferAlleleClusters</span>(</span>
<span id="lst-asc-2"><a href="#lst-asc-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">germline_set =</span> germline, </span>
<span id="lst-asc-3"><a href="#lst-asc-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">trim_3prime_side =</span> <span class="dv">318</span>, </span>
<span id="lst-asc-4"><a href="#lst-asc-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">mask_5prime_side =</span> <span class="dv">0</span>, </span>
<span id="lst-asc-5"><a href="#lst-asc-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">family_threshold =</span> <span class="dv">75</span>, </span>
<span id="lst-asc-6"><a href="#lst-asc-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">allele_cluster_threshold =</span> <span class="dv">95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output of <code>inferAlleleClusters</code> is an <code>S4</code> object of type <code>GermlineCluster</code> that contains several slots:</p>
<table class="table">
<colgroup>
<col style="width: 25%">
<col style="width: 75%">
</colgroup>
<thead>
<tr class="header">
<th>Slot</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>germlineSet</td>
<td>The input germline set with the 3’ and 5’ modifications (If defined)</td>
</tr>
<tr class="even">
<td>alleleClusterSet</td>
<td>The input germline set with the ASC name scheme, if exists without duplicated sequences</td>
</tr>
<tr class="odd">
<td>alleleClusterTable</td>
<td>The allele similarity cluster with the new names and the default thresholds</td>
</tr>
<tr class="even">
<td>threshold</td>
<td>The input family and allele cluster similarity thresholds</td>
</tr>
<tr class="odd">
<td>hclustAlleleCluster</td>
<td>Germline set hierarchical clustering, an hclust object</td>
</tr>
</tbody>
</table>
<p>We can use the S4 plot method to plot the hierarchical clustering of the germline set as seen below in <a href="#fig-plot">Figure&nbsp;<span>1</span></a>.</p>
<div class="cell" data-crop="true">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(asc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-plot" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="piglet_package_files/figure-html/fig-plot-1.png" class="img-fluid figure-img" width="1440"></p>
<p></p><figcaption class="figure-caption">Fig 1: The out most circle is the allele names, the second layer are the ASC groups, each group is labeled and colored. The third circle is the clustering dendrogram, the braches are colored by the ASC families. The blue and orange dashed lines are the 95% and 75% similarity ASC threshold.</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="artificial-framework-1-reference-set" class="level3">
<h3 class="anchored" data-anchor-id="artificial-framework-1-reference-set">Artificial framework 1 reference set</h3>
<p>As described in <a href="alleles_groups.html#sec-amplicon_Length"><span>Section&nbsp;1</span></a>, not all repertoires data available covers the entire V region. Hence, a modified reference set for the sequenced region can help us further understand the results we can obtain from certain library protocols.</p>
<p>Hence, we created the function <code>artificialFRW1Germline</code>, to mimic the seen coding region of targeted framework 1 (FRW1) primers for a given reference set. The primers were obtained from BIOMED-2 protocol <span class="citation" data-cites="van2003design">(<a href="#ref-van2003design" role="doc-biblioref">Van Dongen et al. 2003</a>)</span>.</p>
<p>Essentially the function matches the primer to each of the germline set sequences and either mask or trim the region. The returned object is a character vector with the named sequence in the desire length (Trimmed/Masked).</p>
<p>To demonstrate the use of the function, we can use the cleaned germline set from above (<a href="#lst-germlineset">Listing&nbsp;<span>1</span></a>). In this case we will mask the FRW1 region, this will return the sequences with the Ns instead of DNA nucleotide. The function output a log of the process, this output can be repressed using the <code>quite=TRUE</code> flag.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>germline_frw1 <span class="ot">&lt;-</span> <span class="fu">artificialFRW1Germline</span>(germline, <span class="at">mask_primer =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>282/286 germline sequences have passed
Counts by primers: 
VH1-FR1:53,VH2-FR1:25,VH3-FR1:122,VH4-FR1:69,VH5-FR1:10,VH6-FR1:3</code></pre>
</div>
</div>
<p>We will look at one sequence, “IGHV1-8*01” to see the masking:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>allele <span class="ot">=</span> <span class="st">"IGHV1-8*01"</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>dna_seq <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">DNAStringSet</span>(<span class="fu">setNames</span>(<span class="fu">c</span>(germline[allele], germline_frw1[allele]), </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                                             <span class="fu">c</span>(<span class="st">"IGHS1V1-8*01"</span>, <span class="st">"IGHS2V1-8*01"</span>)))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">names =</span> <span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"T"</span>,<span class="st">"C"</span>,<span class="st">"G"</span>,<span class="st">"N"</span>,<span class="st">"."</span>), </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">color =</span> <span class="fu">c</span>(<span class="st">"#ff6d6d"</span>,<span class="st">"#769dcc"</span>,<span class="st">"#f2be3c"</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"#74ce98"</span>,<span class="st">"#b8b8b8"</span>,<span class="st">"#ffffff"</span>))</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>ggmsa<span class="sc">::</span><span class="fu">ggmsa</span>(dna_seq, <span class="at">custom_color =</span> colors, <span class="at">show.legend =</span> T) <span class="sc">+</span> </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  ggmsa<span class="sc">::</span><span class="fu">facet_msa</span>(<span class="at">field =</span> <span class="dv">80</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="piglet_package_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can use the artificial germline set to infer the ASC clusters in the same fashion as in <a href="#sec-ASC"><span>Section&nbsp;2.1</span></a></p>
</section>
</section>
<section id="allele-based-genotype" class="level1">
<h1>Allele based genotype</h1>
<p>Genotype inference is important for downstream analysis of B cell repertoires. Utilizing this technique can reduce bias annotating the repertoire, gene and allele usage analysis, and much more (<a href="allele_based_genotype.html"><span>Chapter&nbsp;1</span></a>). In PIgLET the Allele based genotype section is dedicated to the ASC-based genotype inference.</p>
<table class="table">
<colgroup>
<col style="width: 27%">
<col style="width: 72%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>inferGenotypeAllele</td>
<td>Infer the IGHV genotype using the allele based method</td>
</tr>
<tr class="even">
<td>assignAlleleClusters</td>
<td>Renames the v allele calls based on the new allele clusters</td>
</tr>
<tr class="odd">
<td>germlineASC</td>
<td>Converts IGHV germline set to ASC germline set</td>
</tr>
<tr class="even">
<td>recentAlleleClusters</td>
<td>Download the most recent version of the allele clusters table archive from zenodo</td>
</tr>
<tr class="odd">
<td>extractASCTable</td>
<td>Extracts the allele cluster table from the zenodo archive file</td>
</tr>
</tbody>
</table>
<section id="asc-based-thresholds" class="level3">
<h3 class="anchored" data-anchor-id="asc-based-thresholds">ASC-based thresholds</h3>
<p>Part of the IGHV reference book is to explore the distribution of the alleles in the population and determine the appropriate allele threshold. The default threshold for any allele is <span class="math inline">\(0.0001\)</span>, this value is also what the function <code>inferAlleleClusters</code> returns for each of the alleles in the germline set. To use the threshold in the reference book (<a href="allele_based_genotype.html#sec-threshold"><span>Section&nbsp;1.1</span></a>), PIgLET include two functions to retrieve the latest update thresholds from the zenodo archive (doi:)</p>
<p>First we will retrieve the archive files using the <code>recentAlleleClusters</code> function. The function can get a path value for locally saving the archive files with the <code>path</code> flag, if non is supplied then the function save the files in a temporary directory. The flag <code>get_file=TRUE</code>, will return the downloaded file full path.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>asc_archive <span class="ot">&lt;-</span> <span class="fu">recentAlleleClusters</span>(<span class="at">doi=</span><span class="st">"10.5281/zenodo.7401189"</span>, <span class="at">get_file =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To extract the ASC threshold table we can use the <code>extractASCTable</code> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>allele_cluster_table <span class="ot">&lt;-</span> <span class="fu">extractASCTable</span>(<span class="at">archive_file =</span> asc_archive)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The table is has identical ASC clusters to the table we created above (<a href="#lst-asc">Listing&nbsp;<span>2</span></a>).</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  new_allele   imgt_allele.piglet imgt_allele.zenodo
  &lt;chr&gt;        &lt;chr&gt;              &lt;chr&gt;             
1 IGHVF1-G1*01 IGHV3-72*01        IGHV3-72*01       
2 IGHVF1-G2*01 IGHV3-73*01        IGHV3-73*01       
3 IGHVF1-G2*02 IGHV3-73*02        IGHV3-73*02       
4 IGHVF1-G3*01 IGHV3-49*02        IGHV3-49*02       
5 IGHVF1-G3*02 IGHV3-49*01        IGHV3-49*01       
6 IGHVF1-G3*03 IGHV3-49*05        IGHV3-49*05       </code></pre>
</div>
</div>
<p>We can now extract the threshold from the zenodo archive table and fill the table created using the PIgLET. We recommend that in case an allele does not have a threshold in the archive to keep the default threshold of <span class="math inline">\(0.0001\)</span>.</p>
</section>
<section id="inferring-asc-based-genotype" class="level3">
<h3 class="anchored" data-anchor-id="inferring-asc-based-genotype">Inferring ASC-based genotype</h3>
<p>Genotype inference has an increasing importance in downstream analysis, as described in <a href="allele_based_genotype.html"><span>Chapter&nbsp;1</span></a> an individual genotype inference can help reduce bias within the repertoire annotations. Based on the reference book, the ASC clusters, and the ASC-based threshold we developed in PIgLET a genotype inference function which is based on the ASC-based genotype.</p>
<p>The function <code>inferGenotypeAllele</code> infer an subject genotype using the absolute fraction and the allele based threshold. Essentially, for each unique allele that is found in the repertoire, its absolute fraction is calculated and compared to the population derived threshold. In case the allele’s fraction is above the threshold then it is inferred into the subject genotype.</p>
<p>Recommendations:</p>
<ul>
<li>For naive repertoires:
<ul>
<li>Filter the repertoire for up to 3 mutation within the V region</li>
<li>Setting the flag <code>find_unmutated=T</code>. Not needed if the above mutation filter is applied</li>
</ul></li>
<li>For non-naive repertoires:
<ul>
<li>Cloning the repertoire and selecting a single clonal representative with the least amount of mutations</li>
<li>Setting the flag <code>single_assignment=F</code>. In this case the function treats cases of multiple allele call assignment as belonging to all groups.</li>
</ul></li>
</ul>
<p>Below is a demonstration of inferring the genotype for an example dataset taken from TIgGER<span class="citation" data-cites="TIgGER">(<a href="#ref-TIgGER" role="doc-biblioref">Gadala-Maria et al. 2015</a>)</span> package.</p>
<p>The data is b cell repertoire data from individual (PGP1) in AIRR format. The records were annotated with by IMGT/HighV-QUEST.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># loading TIgGER AIRR-seq b cell data</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> tigger<span class="sc">::</span>AIRRDb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For using the genotype inference function on non ASC name scheme annotations, we first need to transform the <code>v_call</code> column to the ASC alleles. We will use the ASC-table downloaded from zenodo archive and the example data</p>
<p>First we will collapse allele duplication in the ASC-table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>allele_cluster_table <span class="ot">&lt;-</span> allele_cluster_table <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(new_allele, func_group, thresh) <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                        dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">imgt_allele =</span> <span class="fu">paste0</span>(<span class="fu">sort</span>(<span class="fu">unique</span>(imgt_allele)), <span class="at">collapse =</span> <span class="st">"/"</span>), <span class="at">.groups =</span> <span class="st">"keep"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can transform the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># storing original v_call values</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>v_call_or <span class="ot">&lt;-</span> data<span class="sc">$</span>v_call</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># assigning the ASC alleles</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>asc_data <span class="ot">&lt;-</span> <span class="fu">assignAlleleClusters</span>(data, allele_cluster_table)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(asc_data[,<span class="fu">c</span>(<span class="st">"v_call"</span>,<span class="st">"v_call_or"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  v_call                      v_call_or              
  &lt;chr&gt;                       &lt;chr&gt;                  
1 IGHVF5-G29*03               IGHV1-2*02             
2 IGHVF5-G30*02               IGHV1-18*01            
3 IGHVF5-G26*10               IGHV1-69*06            
4 IGHVF5-G26*07               IGHV1-69*04            
5 IGHVF5-G27*02               IGHV1-8*01             
6 IGHVF5-G28*03,IGHVF5-G28*02 IGHV1-46*01,IGHV1-46*03</code></pre>
</div>
</div>
<p>If we have not inferred the ASC clustered and generated the renamed germline set, we can use the <code>germlineASC</code> to obtain it. We need to supply the function the ASC-table and an IGHV germline set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reforming the germline set</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>asc_germline <span class="ot">&lt;-</span> <span class="fu">germlineASC</span>(allele_cluster_table, <span class="at">germline =</span> HVGERM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once we have both the modified dataset and germline reference set, we can infer the genotype. The function returns the genotype table with the following columns</p>
<table class="table">
<colgroup>
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th>gene</th>
<th>alleles</th>
<th>imgt_alleles</th>
<th>counts</th>
<th>absolute_fraction</th>
<th>absolute_threshold</th>
<th>genotyped_alleles</th>
<th>genotype_imgt_alleles</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>allele cluster</td>
<td>the present alleles</td>
<td>the imgt nomenclature</td>
<td>the number of reads</td>
<td>the absolute fraction</td>
<td>the population driven allele</td>
<td>the alleles which</td>
<td>the imgt nomenclature</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inferring the genotype</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>asc_genotype <span class="ot">&lt;-</span> <span class="fu">inferGenotypeAllele</span>(asc_data, </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="at">alleleClusterTable =</span> allele_cluster_table, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="at">germline_db =</span> asc_germline, <span class="at">find_unmutated=</span>T)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(asc_genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         gene     alleles
1: IGHVF5-G22          01
2: IGHVF5-G23          01
3: IGHVF5-G24       02,03
4: IGHVF5-G26 15,07,10,01
5: IGHVF5-G27          02
6: IGHVF5-G28          03
                                                   imgt_alleles        counts
1:                                                  IGHV1-24*01           105
2:                                                IGHV1-69-2*01            31
3:                                      IGHV1-58*01,IGHV1-58*02         23,18
4: IGHV1-69*01/IGHV1-69D*01,IGHV1-69*04,IGHV1-69*06,IGHV1-69*02 515,469,280,9
5:                                                   IGHV1-8*01           467
6:                                                  IGHV1-46*01           624
                         absolute_fraction          absolute_threshold
1:                               0.0221613                      0.0001
2:                               0.0065429                      0.0001
3:                     0.0048544,0.0037991               0.0001,0.0001
4: 0.1086956,0.0989869,0.0590967,0.0018995 0.0010,0.0010,0.0010,0.0010
5:                               0.0985648                      0.0001
6:                               0.1317011                      0.0010
   genotyped_alleles
1:                01
2:                01
3:             02,03
4:       15,07,10,01
5:                02
6:                03
                                         genotyped_imgt_alleles
1:                                                  IGHV1-24*01
2:                                                IGHV1-69-2*01
3:                                      IGHV1-58*01,IGHV1-58*02
4: IGHV1-69*01/IGHV1-69D*01,IGHV1-69*04,IGHV1-69*06,IGHV1-69*02
5:                                                   IGHV1-8*01
6:                                                  IGHV1-46*01</code></pre>
</div>
</div>
<p>For plotting the genotype with TIgGER <code>plotGenotype</code>, we need to do a small modification to our genotype table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the genotype alleles</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>alleles <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(asc_genotype<span class="sc">$</span>genotyped_imgt_alleles,<span class="st">","</span>))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># get the genes</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"[*][0-9]+"</span>,<span class="st">""</span>,alleles)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the alleles</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>alleles <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(alleles,<span class="st">"[*]"</span>),<span class="st">"[["</span>,<span class="dv">2</span>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure to extract only alleles</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>alleles <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"([0-9]+).*$"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, alleles)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co"># create the genotype</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>genotype <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">alleles =</span> alleles, <span class="at">gene =</span> genes)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the genotype</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>tigger<span class="sc">::</span><span class="fu">plotGenotype</span>(<span class="at">genotype =</span> genotype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="piglet_package_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-TIgGER" class="csl-entry" role="doc-biblioentry">
Gadala-Maria, Daniel, Gur Yaari, Mohamed Uduman, and Steven H. Kleinstein. 2015. <span>“Automated Analysis of High-Throughput b-Cell Sequencing Data Reveals a High Frequency of Novel Immunoglobulin v Gene Segment Alleles.”</span> <em>Proceedings of the National Academy of Sciences</em> 112 (8): E862–70. <a href="https://doi.org/10.1073/pnas.1417683112">https://doi.org/10.1073/pnas.1417683112</a>.
</div>
<div id="ref-van2003design" class="csl-entry" role="doc-biblioentry">
Van Dongen, JJM, AW Langerak, M Brüggemann, PAS Evans, M Hummel, FL Lavender, E Delabesse, et al. 2003. <span>“Design and Standardization of PCR Primers and Protocols for Detection of Clonal Immunoglobulin and t-Cell Receptor Gene Recombinations in Suspect Lymphoproliferations: Report of the BIOMED-2 Concerted Action Bmh4-Ct98-3936.”</span> <em>Leukemia</em> 17 (12): 2257–317.
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./allele_based_genotype.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">ASC-based genotype {.unnumbered}</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./cluster_docker.html" class="pagination-link">
        <span class="nav-page-text">Processing AIRR-seq data with the ASC-pipeline</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center">© Copyright 2022 of YaariLab (gur.yaari@biu.ac.il), website licensed under the [CC BY-NC-ND, 4.0] (https://creativecommons.org/licenses/by-nc-nd/4.0/)</div>
  </div>
</footer>



</body></html>